adc: stm32: fix die temperature readings

fouge opened the pull request:
ADC readings for die temperature on STM32 are sporadically negatives since commit 588277c43a0e12d176d377ab440c8b7640a6c3a4.

The bug appears consistently on STM32G4. I bisected the commits on `adc_stm32.c` file to find out which commit is causing issue and found a quick way to solve it.

The solution is to re-enable the ADC whenever is has been disabled, but I am not 100% confident this is the best solution cause I don't explain this bug. 

cc @ycsin

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -536,51 +538,68 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
-		k_usleep(LL_ADC_DELAY_VREFINT_STAB_US);
+		if (ret == 0) {
+			adc_stm32_enable(adc);
+#ifdef LL_ADC_DELAY_TEMPSENSOR_STAB_US
```
LL_ADC_DELAY_VREFINT_STAB_US

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -536,51 +538,68 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
+		if (ret == 0) {
+			adc_stm32_enable(adc);
+#ifdef LL_ADC_DELAY_TEMPSENSOR_STAB_US
+			k_usleep(LL_ADC_DELAY_TEMPSENSOR_STAB_US);
```
LL_ADC_DELAY_VREFINT_STAB_US

fouge replied:
fixed

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -536,51 +538,68 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
 	if (config->has_temp_channel) {
 #if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
 		if ((__LL_ADC_CHANNEL_TO_DECIMAL_NB(LL_ADC_CHANNEL_TEMPSENSOR_ADC1) == channel_id)
 		    && (config->base == ADC1)) {
-			adc_stm32_disable(adc);
+			ret = adc_stm32_disable(adc);
 			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-			k_usleep(LL_ADC_DELAY_TEMPSENSOR_STAB_US);
+			if (ret == 0) {
+				adc_stm32_enable(adc);
+				k_usleep(LL_ADC_DELAY_TEMPSENSOR_STAB_US);
+			}
```
As the code section grows, factorizing it would be nice.
(Not blocking though)

fouge commented:
@ycsin it would be great to have your insights on this 🙂

erwango commented:
@gautierg-st Can you review ?

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -537,51 +539,68 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
-			k_usleep(LL_ADC_DELAY_TEMPSENSOR_STAB_US);
+			if (ret == 0) {
+				adc_stm32_enable(adc);
+				k_usleep(LL_ADC_DELAY_TEMPSENSOR_STAB_US);
```
If the adc is already enabled (returned !=0), setting the adc_stm32_set_common_path is still done and it must wait for the LL_ADC_DELAY_TEMPSENSOR_STAB_US  stabilisation. This is not conditionned by the adc enable or not situation. That would be : 
```
			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
			k_usleep(LL_ADC_DELAY_TEMPSENSOR_STAB_US);
			if (ret == 0) {
				adc_stm32_enable(adc);
			}
```

FRASTM replied:
which does not differ from the current code.

fouge replied:
I reverted changes on the disable function and reordered the call to disable -> k_usleep -> enable

FRASTM commented:
In case of the stm32G4, "_The software is allowed to set ADEN only when all bits of ADC_CR registers are 0
(ADCAL=0, JADSTART=0, ADSTART=0, ADSTP=0, ADDIS=0 and ADEN=0) except for
bit ADVREGEN which must be 1 (and the software must have wait for the startup time of
the voltage regulator)_" 
Could this be checked during the adc_stm32_enable()  ?

fouge committed:
adc: stm32: stop ADC before disabling it

It is advised to stop any ongoing ADC conversion before
disabling the peripheral.
Added some comments so that ADC state isn't left into
intermediate state that would prevent correctly
enabling or disabling the peripheral

Signed-off-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

fouge commented:
@FRASTM 
From what I understand if the driver is used correctly, we shouldn't be able to leave the peripheral in a bad state that would prevent enabling it.
Calibration function waits for the ADCAL bit to be reset, disable function waits for ADEN to be reset.

I added some checks in latest commit though to make sure we don't disable the peripheral with any ongoing conversion, like advised in the reference manual (I checked STM32G4 reference manual, likely the same for other series right?)
I also added comment to make sure we don't leave the ADC in intermediate state when enabling/disabling the peripheral and when using calibration function.

erwango commented:
@fouge can you rebase? I'm surprised github doesn't complain, but it required.

fabiobaltieri commented:
@fouge can you rebase please?
